File: library/vision_detection.py
# ################################################################################################
# ------------------------------------------------------------------------------------------------
# File:   vision_detection.py
# Author: Luis Monteiro
#
# Created on nov 8, 2019, 22:00 PM
# ------------------------------------------------------------------------------------------------
# ################################################################################################
#
# external imports
from cv2      import waitKey    as wait
from logging  import getLogger  as logger
#
# internal imports
from library.inputs.vision_input_camera   import VisionInputCamera   as DefaultInput
from library.outputs.vision_output_window import VisionOutputWindow  as DefaultOutput
from library.vision_config                import VisionFilterBuilder
# ################################################################################################
# ------------------------------------------------------------------------------------------------
# VisionDetection 
# ------------------------------------------------------------------------------------------------
# ################################################################################################
class VisionDetection:
    #
    # -----------------------------------------------------------------------------------
    # initialization
    # -----------------------------------------------------------------------------------
    #
    def __init__(self, config, input=DefaultInput(), output=DefaultOutput()):
        # -------------------------------------------------
        # variables
        # -------------------------------------------------
        # command containers  
        self.__select = []
        # load input 
        self.__input  = input
        # load output
        self.__output = output
        # load logger
        self.__logger = logger()
        # filter data base 
        self.__filters = self._load_filters(config)
    #
    # -----------------------------------------------------------------------------------
    # serve
    # -----------------------------------------------------------------------------------
    #
    def serve(self, observer):
        #
        # check vision input
        #
        if not self.__input.good():
            raise RuntimeError("vision input open fail ...")
        # 
        # image processing
        #
        selected = {}
        while wait(1000) < 0:
            # read input frame
            frame = self.__input.read()
            # write output frame
            self.__output.write_frame(frame)
            # update selected filters
            for id, params in self._read_commands():
                if params is not None:
                    filter = self.__filters[id]
                    filter.update(**params)
                    selected[id] = filter
                else:
                    selected.pop(id, None) 
            # process filters
            for id, filter in selected.items():
                try:
                    self.__output.write_filter(
                        id, filter.region(), filter.process(frame))
                except:
                    self.__logger.exception("process filter {} failed".format(id))
            # output flush
            self.__output.flush(observer)
    #
    # -----------------------------------------------------------------------------------
    # write commands
    # -----------------------------------------------------------------------------------
    #
    def set_filter(self, id, **kargs):
        self.__select.append((id, kargs))
        return self
    def clr_filter(self, id):
        self.__select.append((id, None))
        return self
    def set_filters(self, **kargs):
        for id in self.__filters:
            self.__select.append((id, kargs))
        return self
    def clr_filters(self):
        for id in self.__filters:
            self.__select.append((id, None))
        return self
    #
    # -----------------------------------------------------------------------------------
    # read commands
    # -----------------------------------------------------------------------------------
    #
    def _read_commands(self):
        return self.__select
    #
    # -----------------------------------------------------------------------------------
    # load filters
    # -----------------------------------------------------------------------------------
    #
    def _load_filters(self, conf):
        filters = {}
        for filter_key, filter_conf in conf['filters'].items():
            try:
                filters[filter_key] = VisionFilterBuilder.Build(
                    filter_conf['type'],
                    filter_conf.get('conf'))
            except Exception as ex:
                self.__log.exception(ex)
        return filters
# #################################################################################################
# -------------------------------------------------------------------------------------------------
# End
# -------------------------------------------------------------------------------------------------
# #################################################################################################

File: library/vision_detection.cpp
/**
 * ------------------------------------------------------------------------------------------------
 * File:   STime.cpp
 * Author: Luis Monteiro
 * 
 * Created on Mai 19, 2019, 22:29 PM
 * ------------------------------------------------------------------------------------------------
 **
 * std
 */
#include <iostream>
#include <fstream>
#include <string>  
#include <iomanip>
#include <sstream>
/**
 * opencv
 */
#include <opencv2/videoio.hpp>
#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>
/**
 * local
 */
#include "vision_detection.hpp"
#include "vision_config.hpp"
#include "vision_video_display.hpp"
/**
 * namespaces
 */
using namespace std;
using namespace cv;
/**
 * ----------------------------------------------------------------------------
 * serve
 * ----------------------------------------------------------------------------
 */
int VisionDetection::serve(function<void(FilterID, VisionFilter::SharedResult)> callback) {				
	VideoDisplay display("1");
	VideoCapture capture(0);
	/**
	 * setup video capture
	 */
	capture.set(CAP_PROP_FRAME_WIDTH, 1280);
	capture.set(CAP_PROP_FRAME_HEIGHT, 960);
	capture.set(CAP_PROP_AUTOFOCUS, 0);
	if (!capture.isOpened()){
		cout << "Could not open reference " << endl;
		return -1;
	}
	/**
	 * image processing
	 */
	for(VideoFrame frame; capture.read(frame); display.print(frame)) {
		/**
		 * select filters
		 */
		for(auto& cmd : _read_commands()) {
			if(std::get<1>(cmd) > 0) {
				__selected.emplace(std::get<0>(cmd), std::get<1>(cmd));
			} else {
				__selected.erase(std::get<0>(cmd));
			}
		}
		/**
		 * process filters
		 */
		for(auto& s: __selected) {
			try {
				for(auto& result: __filters.at(std::get<0>(s))->process(frame)) {
					if(result->level() > std::get<1>(s)) {
						callback(std::get<0>(s), result);
					}
				}
			} catch(...) {

			}
		}
		if (waitKey(10) == 27) {
			cout << "Esc key is pressed by user. Stoppig the video" << endl;
			return 0;
		}
	}	
	cout << "Found the end of the video" << endl;
	return 0;
}
/**
 * ----------------------------------------------------------------------------
 * load filters
 * ----------------------------------------------------------------------------
 */
VisionDetection::Filters VisionDetection::_load_filters(std::istream& input) {
	auto conf = FilterConfig::Load(input);
	/**
	 * load filters
	 */
	auto filters = VisionDetection::Filters();
	for(auto filter: conf["filters"]) {
		try {
			filters.emplace(
				std::get<0>(filter).as<FilterID>(), 
				FilterBuilder::Build(
					std::get<1>(filter)["type"].as<string>(), 
					std::get<1>(filter)["conf"]
				)
			);
		} catch(std::exception&) {
			Log::Warning("fail to load {} filter", std::get<0>(filter).as<FilterID>());
		}
	}
	return filters;
}
/**
 * ------------------------------------------------------------------------------------------------
 * End
 * ------------------------------------------------------------------------------------------------
 */
